### ISSUE TITLE
New rule: H33 - link-text

### ISSUE THREAD

**Issue by jmucha** [dequelabs/axe-core#537]
#  H33 - link-text

The objective of this test is to check whether an additional text (`title` element) is provided to describe the purpose of the link.

**Tags:** wcag2a, wcag244, cat.text-alternatives, best-practice

## Selector

Select all anchor `a` elements.

### text-alternatives/link-text (any)

1.    For each `a` element check if the link text is unique. If true, pass; else 2.
2.    Check if the links have different titles. If true, pass; else fail.

**Comment by WilcoFiers** [comment]
I don't think this one works all that well. Links aren't required to have a title by WCAG 2.4.4. What I'd suggest doing with this one is to look for links with the same link text, and check if they have different titles. This would be a best-practice.

**Comment by devarshipant** [comment]
If title is present on a link, at a minimum, the test should check and report if the title contains part of the link text. The tester should then manually check the context in which the title is used. Manual override is required to make a determination without calling it a violation.

If the title is exactly the same as the link (innerText), then JAWS 17 / 18 ignore the title and NVDA does not bother speaking the title, so that can be considered a pass.

**Comment by jmucha** [comment]
Let's do it as you suggested @WilcoFiers

**Comment by Danidude** [comment]
Looking over this rule, should the selector be a[title]? As if the link do not contain an title it is no point in checking if the text is different?

**Comment by WilcoFiers** [comment]
It's not just about the title. It's about the link text + title. I would suggest using `a[href]`. We don't want to catch anchor elements without href in this.

**Comment by Danidude** [comment]
Okay, will make this change then, then it also should be ´[role=link]´?

**Comment by Danidude** [comment]
An few more questions are as follows.

 - This test should also compare the href value, to see if they are different. Example here on github the "Pricing" link is the same text as bottom and top, and same URL. But it should not fail an test as this. So link with same text and same href should also pass. 

 - Also what defines 'link text'? Is it only the text between the `<a>` tag, or is aria-label/labelledby also included?

 - This rule looks a lot similar to an other rule we are writing: #532 . Should one of these be dropped? Or re-written?

**Comment by jkodu** [comment]
Hi Dani,

Answers to your questions below:

<blockquote>
Okay, will make this change then, then it also should be [role=link]?
</blockquote>
A: 
An anchor element with a `href` will have a semantic `role` of `link`. But `[role='link']` can perhaps be set on other element types like `button`. Believe this rule is catering only to anchor elements. 
Something to note is that you could write a `{rule-id}-matches.js` to filter out elements/ node with the right attributes etc, rather than try and come up with a complex selector in the  rule spec json file.

<blockquote> 
This test should also compare the href value, to see if they are different. Example here on github the "Pricing" link is the same text as bottom and top, and same URL. But it should not fail an test as this. So link with same text and same href should also pass. 
</blockquote>
A: Yes, I agree with you. As @WilcoFiers mentions, this rule identifies links with the same link text, and check if they have different titles. This would be a best-practice.

<blockquote> 
Also what defines 'link text'? Is it only the text between the <a> tag, or is aria-label/labelledby also included?
</blockquote>
A: I am not sure on this one Dani. But I think it is just the `link text` and not the `aria-*` attributes.

<blockquote> 
This rule looks a lot similar to an other rule we are writing: #532. Should one of these be dropped? Or re-written?
</blockquote>
A: I do not think they both are the same rules, I will leave my notes on the other one.

**Comment by WilcoFiers** [comment]
This proposal is superseded by https://github.com/dequelabs/axe-core/issues/1505.
